File: task-manager/python/src/app/models/task.py
from datetime import datetime
from enum import StrEnum
from uuid import UUID

from pydantic import BaseModel


class TaskStatus(StrEnum):
    ToDo = "todo"
    InProgress = "in_progress"
    Done = "done"


class BaseTask(BaseModel):
    title: str
    description: str | None = None
    status: TaskStatus = TaskStatus.ToDo
    due_date: datetime | None = None


class CreateTask(BaseTask):
    pass


class UpdateTask(BaseModel):
    title: str | None = None
    description: str | None = None
    status: TaskStatus | None = None
    due_date: datetime | None = None


class ReadTask(BaseTask):
    pk: UUID
    user: UUID
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True

File: task-manager/typescript/src/dto/task.ts
import { PartialType } from "@nestjs/swagger";
import { IsDateString, IsEnum, IsOptional, IsString } from "class-validator";

export enum TaskStatus {
  TODO = 'todo',
  IN_PROGRESS = 'in_progress',
  DONE = 'done',
}

export class CreateTaskDto {
    @IsString()
    title: string;

    @IsString()
    @IsOptional()
    description?: string;

    @IsEnum(TaskStatus)
    @IsOptional()
    status?: TaskStatus;

    @IsDateString()
    @IsOptional()
    due_date?: string;
}

export class UpdateTaskDto extends PartialType(CreateTaskDto) {}